src/js/crud.js

```diff
@@ -0,0 +1,168 @@
+  const bases = await getBaseUrls();
+  return `${bases[0]}/${endpoint}`;
+}
+
+// Predefined tags
+const predefinedTags = [
+  "bread", "artisan", "savory", "vegan", "special!", 
+  "pastry", "buttery", "breakfast", "sweet", "dessert",
+  "chocolate", "fruit", "snack", "cake", "citrus"
+];
+
+// Helpers
+function highlightError(inputEl) {
+  inputEl.style.borderColor = "red";
+  inputEl.focus();
+}
+function clearErrors(inputs) {
+  inputs.forEach(i => i.style.borderColor = "#ccc");
+}
+
+// Populate tags dropdown
+function populateTagsDropdown() {
+  const tagsSelect = document.getElementById("tags");
+  predefinedTags.forEach(tag => {
+    const option = document.createElement("option");
+    option.value = tag;
+    option.textContent = tag;
+    tagsSelect.appendChild(option);
+  });
+}
+
+// Fetch and render products
+async function fetchProducts() {
+  const bases = await getBaseUrls();
+  const products = await fetchFromMultipleBases('products', bases);
+
+  const list = document.querySelector('.product-list');
+  list.innerHTML = ""; // clear old list
+
+  products.forEach(p => {
+    const li = document.createElement("li");
+
+    const h3 = document.createElement("h3");
+    h3.textContent = p.name;
+
+    const price = document.createElement("p");
+    price.textContent = `Price: €${p.price.toFixed(2)}`;
+
+    const desc = document.createElement("p");
+    if (p.description) desc.textContent = p.description;
+
+    const img = document.createElement("img");
+    if (p.image) {
+      img.src = p.image;
+      img.alt = p.name;
+    }
+
+    const tagsP = document.createElement("p");
+    tagsP.className = "tags";
+    if (p.tags) {
+      tagsP.textContent = Array.isArray(p.tags) ? p.tags.join(", ") : p.tags;
+    }
+
+    const deleteBtn = document.createElement("button");
+    deleteBtn.className = "btn delete-btn";
+    deleteBtn.textContent = "Delete";
+    deleteBtn.addEventListener("click", async () => {
+      if (!confirm("Are you sure you want to delete this product?")) return;
+      const url = await getBaseUrl('products');
+      await fetch(`${url}/${p.id}`, { method: "DELETE" });
+      fetchProducts();
+    });
+
+    const editBtn = document.createElement("button");
+    editBtn.className = "btn edit-btn";
+    editBtn.textContent = "Edit";
+    editBtn.addEventListener("click", async () => {
+      const updatedProduct = {
+        name: prompt("Name:", p.name),
+        price: parseFloat(prompt("Price:", p.price)),
+        image: prompt("Image URL:", p.image),
+        description: prompt("Description:", p.description),
+        tags: prompt("Tags (comma separated):", p.tags).split(",").map(t => t.trim())
+      };
+
+      if (!updatedProduct.name) return alert("Name is required!");
+      if (isNaN(updatedProduct.price) || updatedProduct.price <= 0) return alert("Price must be positive!");
+
+      const url = await getBaseUrl('products');
+      await fetch(`${url}/${p.id}`, {
```

Comment:
Voordat je de fetch verstuurd, check of de `url` wel een waarde bevat. Anders krijg je vervelende errors